### PR TITLE
o/snapstate: disk space check with InstallMany

### PR DESCRIPTION
**Pull request by stolowski**
Use disk space check introduced with InstallMany.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
nice!

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -861,6 +861,29 @@ func InstallMany(st *state.State, names []string, userID int) ([]string, []*stat
...
+	totalSize, err := installSize(st, snapInfos, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+	// require 5Mb extra
+	requiredSpace := totalSize + 5*1024*1024
+	path := dirs.SnapdStateDir(dirs.GlobalRootDir)
+	if err := osutilCheckFreeSpace(path, requiredSpace); err != nil {
+		if _, ok := err.(*osutil.NotEnoughDiskSpaceError); ok {
+			return nil, nil, &ErrInsufficientSpace{
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:879]: this should be called InsufficientSpaceError per conventions, also it should probably carray a ChangeKind like ChangeConflictError

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -861,6 +861,29 @@ func InstallMany(st *state.State, names []string, userID int) ([]string, []*stat
...
+	snapInfos := make([]*snap.Info, len(installs))
+	for i, sar := range installs {
+		snapInfos[i] = sar.Info
+	}
+	totalSize, err := installSize(st, snapInfos, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+	// require 5Mb extra
+	requiredSpace := totalSize + 5*1024*1024
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:875]: we already have that number in the remove code right? maybe this should be a small helper  safetyMarginDiskSpace(size)  ?

**Review by pedronis — COMMENTED** [review]
some comments

File: overlord/snapstate/snapstate_install_test.go [hunk 3]
```diff
@@ -3144,6 +3149,20 @@ func (s *snapmgrTestSuite) TestInstallMany(c *C) {
...
+	restore := snapstate.MockOsutilCheckFreeSpace(func(string, uint64) error { return &osutil.NotEnoughDiskSpaceError{} })
+	defer restore()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	_, _, err := snapstate.InstallMany(s.state, []string{"one", "two"}, 0)
+	diskSpaceErr := err.(*snapstate.InsufficientSpaceError)
+	c.Assert(diskSpaceErr, ErrorMatches, `insufficient space in .* to perform "install" change for the following snaps: one, two`)
+	c.Check(diskSpaceErr.Path, Equals, filepath.Join(dirs.GlobalRootDir, "/var/lib/snapd"))
```
Comment (pedronis) [inline overlord/snapstate/snapstate_install_test.go:3162]: add a check for ChangeKind?

File: overlord/snapstate/snapstate_test.go [hunk 4]
```diff
@@ -1089,7 +1089,7 @@ func (s *snapmgrTestSuite) TestRemoveDiskSpaceForSnapshotError(c *C) {
 	})
 
 	_, err := snapstate.Remove(s.state, "some-snap", snap.R(0), nil)
-	diskSpaceErr := err.(*snapstate.ErrInsufficientSpace)
+	diskSpaceErr := err.(*snapstate.InsufficientSpaceError)
 	c.Assert(diskSpaceErr, ErrorMatches, `cannot create automatic snapshot when removing last revision of the snap: insufficient space.*`)
 	c.Check(diskSpaceErr.Path, Equals, filepath.Join(dirs.GlobalRootDir, "/var/lib/snapd"))
```
Comment (pedronis) [inline overlord/snapstate/snapstate_test.go:1094]: same

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -74,26 +74,33 @@ var ErrNothingToDo = errors.New("nothing to do")
 
 var osutilCheckFreeSpace = osutil.CheckFreeSpace
 
-type ErrInsufficientSpace struct {
+type InsufficientSpaceError struct {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:77]: // InsufficientSpaceError represent an error where the is not enough disk space to perform an operation.

File: overlord/snapstate/snapstate.go [hunk 6]
```diff
@@ -74,26 +74,33 @@ var ErrNothingToDo = errors.New("nothing to do")
...
 	}
 	if len(e.Snaps) > 0 {
 		snaps := strings.Join(e.Snaps, ", ")
-		return fmt.Sprintf("insufficient space in %q to perform operation for the following snaps: %s", e.Path, snaps)
+		return fmt.Sprintf("insufficient space in %q to perform %q change for the following snaps: %s", e.Path, e.ChangeKind, snaps)
 	}
 	return fmt.Sprintf("insufficient space in %q", e.Path)
 }
 
+func safetyMarginDiskSpace(size uint64) uint64 {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:99]: // safetyMarginDiskSpace returns size plus a safety margin (5Mb)

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -74,26 +74,33 @@ var ErrNothingToDo = errors.New("nothing to do")
...
+	// 5Mb margin
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:100]: then the comment inside can go away

**Review by pedronis — APPROVED** [review]
thank you, some minor comments

**Pull request merged**